cloudfoundry/resource_cf_app.go

```diff
@@ -649,8 +663,16 @@ func resourceAppUpdate(d *schema.ResourceData, meta interface{}) (err error) {
 			return
 		}
 	}
-	if d.HasChange("stopped") {
-
+	if restart {
```

Comment:
I think some of the complexity that this change adds to the code stems from my, probably misguided, attempt to make my change as discrete as possible due my not currently being able to be up-to-date with the dev branch as well as my current unfamiliarity with the code and the design goals of the team working on it.  After examining this code further, I believe a complete redesign, of how resourceAppUpdate handles changes that require restaging and/or changes to the running state of the application, is justified.  

I propose the following flow for resourceAppUpdate:
1. use am.UpdateApp() in order to configure all of the details
   * can set `restage` flag
1. update service bindings
   * sets `restage`flag if executed
1. update routes
1. update the binary
   * sets `binaryUpdated` flag if executed
1. Here is pseudocode for the new restage/restart logic
   ```go
   restart := false
   curApp = am.ReadApp(app.ID)
   if binaryUpdated {
      if *curApp.PackageState != "PENDING" {
         // if it's not already pending, we need to restage
         restage = true        
      } else {
         // uploading the binary flagged the app for restaging, 
         // but we need to restart in order to force that to happen now
         restage = false
         restart = true
   }
   if restage {
      am.RestageApp(app.ID, timeout)
   } else if restart && !d.Get("stopped").(bool) { // only run restart if the final state is running
      am.StopApp(app.ID, timeout)
      am.StartApp(app.ID, timeout)
   }
   
   // now set the final started/stopped state, whatever it is
   if d.HasChange("stopped") {
      if d.Get("stopped").(bool) {
         am.StopApp(app.ID, timeout)
      } else {
         am.StartApp(app.ID, timeout)
      }
   }
   ```

None of the current code in this change came from the CF CLI, and while its behaviour is a good reference, I don't want to use its code because its equivalent operation (cf push) doesn't actually allow for an application's binary to be updated while leaving the application in a stopped state.

Reply:
Anyone care to comment on the above before I go ahead with the implementation?

@bgandon

Reply:
Yep, sorry for the delay!

Your pseudo code looks great. That's indeed what I meant in my comment.
For the 5-steps process, I couldn't give a valuable advice because I'm not aware enough about how this is supposed to be done. But I trust you do.

So I'm encouraging you in going ahead with this implementation. The other possibility would be to delay this to v1.1 and just quick-fix here for v1.0, but as the App Push feature is not even working in v0.9, I'm in favor of doing it right as soon as we can in order not increase the technical debt that looks already big.